test/Conversion/TritonToLinalg/cat.mlir

```diff
@@ -0,0 +1,44 @@
+// RUN: triton-shared-opt --triton-to-linalg %s | FileCheck %s
+
+module {
+  tt.func public @kernel(%arg0: !tt.ptr<f64> {tt.divisibility = 16 : i32}, %arg1: !tt.ptr<f64> {tt.divisibility = 16 : i32}, %arg2: !tt.ptr<f64> {tt.divisibility = 16 : i32}) attributes {noinline = false} {
+    %0 = tt.make_range {end = 128 : i32, start = 0 : i32} : tensor<128xi32>
+    %1 = tt.splat %arg0 : !tt.ptr<f64> -> tensor<128x!tt.ptr<f64>>
+    %2 = tt.addptr %1, %0 : tensor<128x!tt.ptr<f64>>, tensor<128xi32>
+    %3 = tt.load %2 : tensor<128x!tt.ptr<f64>>
+    %4 = tt.splat %arg1 : !tt.ptr<f64> -> tensor<128x!tt.ptr<f64>>
+    %5 = tt.addptr %4, %0 : tensor<128x!tt.ptr<f64>>, tensor<128xi32>
+    %6 = tt.load %5 : tensor<128x!tt.ptr<f64>>
+    %7 = tt.cat %3, %6 : tensor<128xf64> -> tensor<256xf64>
+    %8 = tt.make_range {end = 256 : i32, start = 0 : i32} : tensor<256xi32>
+    %9 = tt.splat %arg2 : !tt.ptr<f64> -> tensor<256x!tt.ptr<f64>>
+    %10 = tt.addptr %9, %8 : tensor<256x!tt.ptr<f64>>, tensor<256xi32>
+    tt.store %10, %7 : tensor<256x!tt.ptr<f64>>
+    tt.return
+  }
+}
+
+// CHECK: module {
+// CHECK:   func.func @kernel(%arg0: memref<*xf64> {tt.divisibility = 16 : i32}, %arg1: memref<*xf64> {tt.divisibility = 16 : i32}, %arg2: memref<*xf64> {tt.divisibility = 16 : i32}, %arg3: i32, %arg4: i32, %arg5: i32, %arg6: i32, %arg7: i32, %arg8: i32) {
+// CHECK:     [[REINTERPRET_CAST_:%.+]] = memref.reinterpret_cast %arg0 to offset: [0], sizes: [128], strides: [1]{{.*}} : memref<*xf64> to memref<128xf64, strided<[1]>>
+// CHECK:     [[ALLOC_:%.+]] = memref.alloc() : memref<128xf64>
+// CHECK:     memref.copy [[REINTERPRET_CAST_]], [[ALLOC_]] : memref<128xf64, strided<[1]>> to memref<128xf64>
+// CHECK:     [[VAR_0_:%.+]] = bufferization.to_tensor [[ALLOC_]] restrict writable : memref<128xf64>
+// CHECK:     [[REINTERPRET_CAST_0_:%.+]] = memref.reinterpret_cast %arg1 to offset: [0], sizes: [128], strides: [1]{{.*}} : memref<*xf64> to memref<128xf64, strided<[1]>>
+// CHECK:     [[ALLOC_1_:%.+]] = memref.alloc() : memref<128xf64>
+// CHECK:     memref.copy [[REINTERPRET_CAST_0_]], [[ALLOC_1_]] : memref<128xf64, strided<[1]>> to memref<128xf64>
+// CHECK:     [[VAR_1_:%.+]] = bufferization.to_tensor [[ALLOC_1_]] restrict writable : memref<128xf64>
+// CHECK:     [[VAR_2_:%.+]] = tensor.empty() : tensor<256xf64>
+// CHECK:     [[VAR_3_:%.+]] = linalg.generic {indexing_maps = [affine_map<(d0) -> (d0)>, affine_map<(d0) -> (-d0 + 127)>], iterator_types = ["parallel"]} ins([[VAR_0_]] : tensor<128xf64>) outs([[VAR_2_]] : tensor<256xf64>) {
+// CHECK:     ^bb0([[IN_:%.+]]: f64, [[OUT_:%.+]]: f64):
```

Comment:
i believe the affine map (`-d0 + 127`) for the first value is not right. shouldn't it be just `d0 -> d0`? say we're concatenating lhs (tensor<128>) with rhs (tensor<128>), the first 128 elements should map directly to the first 128 elements in the output; then other 128 elements map to the latter half (d0 + 128)

Reply:
we cannot do `d0 -> d0` because these two tensors have different number of elements, this would break in runtime, instead I do this trick of 127 - d0 which makes it iterate over the first half of the elements, since the semantics of "cat" does not require to preserver the original ordering, this trick works. 

Summarizing, to concatenate tensor<128> (a) with tensor<128> (b), I need to allocate tensor<256> (c) for results, then copy first half, copy second half. I cannot iterate on a & c with all_to_all affine map because they have different number of elements, I need to cut indexes > 127. Second half is easy to do with d0 + 127, but for the first I do the trick of 127 - d0 which makes it iterate over 0..127. Let me know what you think! Thank you!

Reply:
thank you for the explanation, that makes sense. this makes me a bit uneasy though because we're trying to replicate the triton nvidia compiler's behaviour to make linalg happy.

i spent some time this evening digging around the mlir code base and found that we can actually just use `tensor.concat`. this op also has an integrated rewrite patterns that decomposes to two copies that are more friendly to accelerators (lowers to just a single dma command as opposed to the vectorized loop in the linalg case).

here's the snippet that i used to test:

```mlir
func.func @concat(%arg0: tensor<128xf32>, %arg1 : tensor<128xf32>) -> tensor<256xf32> {
  %1 = tensor.concat dim(0) %arg0, %arg1 : (tensor<128xf32>, tensor<128xf32>) -> tensor<256xf32>
  return %1 : tensor<256xf32>
}

module attributes {transform.with_named_sequence} {
  transform.named_sequence @__transform_main(%root: !transform.any_op {transform.readonly}) {
    %func_op = transform.structured.match ops{["func.func"]} in %root : (!transform.any_op) -> !transform.op<"func.func">
    transform.apply_patterns to %func_op {
      transform.apply_patterns.tensor.decompose_concat
    } : !transform.op<"func.func">
    transform.yield
  }
}
```

the command to lower the `tensor.concat` to two copies:

```
./bin/mlir-opt concat.mlir -transform-interpreter --one-shot-bufferize=allow-return-allocs-from-loops=true
```

Reply:
Interesting! I initially started with tensor.concat, but it wasn't properly bufferized. I was not aware about transformational patterns, good to know! Let me get back to you with the code, I am totally fine to rewrite it, thanks a lot for your help!

Reply:
done, rewritten with built-in transformational patterns, thanks for the hint!